Basecoin refactor

rigelrozanski opened the pull request:
closes #421

jolesbi commented:
which issue is this related to?

rigelrozanski commented:
@jolesbi updated description

ebuchman commented on examples/basecoin/Makefile:
```diff
@@ -1,41 +1,19 @@
 build:
 	go build $(BUILD_FLAGS) -o build/basecoin ./cmd/...
 
-
```
what happened here? Is this because we assume the tool stuff happens at the top level Makefile, so we can keep these smaller?

rigelrozanski replied:
I couldn't see how/why these tools were being used here. we should obviously not have duplicate tools either. (Can revert of course)

jaekwon replied:
The idea with the tools dir is that every root project have it, and all of our common Makefile and build tooling can go in there.

Really it should be a git submodule for each root project.  So, tendermint/tendermint has ./tools, and so does cosmos/cosmos-sdk/examples/basecoin, because both are root projects that require build tools.

So lets revert and use ./tools to handle things like... github.com/tendermint/glide installation, checking for proper golang version like ensuring it's at least version 1.9, etc... and later we can merge them into github.com/tendermint/make-tools and use it across our repos.

ebuchman commented on examples/basecoin/app/app.go:
```diff
@@ -15,22 +14,20 @@ import (
 	*bam.BaseApp
-	router     bam.Router
-	cdc        *wire.Codec
-	multiStore sdk.CommitMultiStore //TODO distinguish this store from *bam.BaseApp.cms <- is this one master?? confused
```
++!

ebuchman commented on examples/basecoin/types/account.go:
```diff
@@ -3,25 +3,60 @@ package types
+	Name     string         `json:"name"`
+	Address  crypto.Address `json:"address"`
+	Coins    sdk.Coins      `json:"coins"`
+	PubKey   cmn.HexBytes   `json:"public_key"`
```
I guess this needs to be rebased on develop where PubKey/Sequence are removed ?

rigelrozanski replied:
addressed

rigelrozanski commented:
Easiest way to view this PR is to start from examples/basecoin/app/app.go and just follow the new process flow of initialization, if desired have the old process flow there simultaneously to view. 

Main changes made:
 - Merged all files from `examples/basecoin/app/` into `app.go` 
 - Initialize Sequentially, rearranged calls, general cleanup of legibility
 - Created `NewAccountMapperSealed()` to create and seal a typical account mapper
 - Created `MountStoresIAVL()` to mount stores simultaneously 
 - Removed Tools (can add back in but didn't see the point to have here)
 - Removed dependancy of Auth on Baseapp (introduced subpackage `baseapp/testtx`)

jaekwon commented on baseapp/testapp.go:
```diff
@@ -51,6 +51,11 @@ func (tapp *TestApp) RunBeginBlock() {
 	return
 }
 
+// kill resources used by basecapp
```
basecapp

jaekwon commented on baseapp/testapp.go:
```diff
@@ -2,8 +2,8 @@ package baseapp
 	abci "github.com/tendermint/abci/types"
-	"github.com/tendermint/go-crypto"
 
+	"github.com/cosmos/cosmos-sdk/baseapp/testtx"
```
I don't see the point of moving this out?

rigelrozanski replied:
There was a point when I was rearranging when circular dependencies where created in baseapp when it references auth.. since then auth is not referenced in baseapp anymore - but I figured I'd leave this in just because it was a bit annoying to deal with when I did. Will move back in if you think it's better.... also maybe `auth` should be adjusted to not include reference to testing whether it's this one `testtx` type... is there another way around that? thoughts?

jaekwon commented on examples/basecoin/app/app.go:
```diff
@@ -1,95 +1,121 @@
+	)
+
+	// add handlers
+	app.Router().AddRoute("bank", bank.NewHandler(bank.NewCoinKeeper(app.accountMapper)))
```
I feel like the coinkeeper should be set to app first.

rigelrozanski replied:
Is this what you meant? 
``` golang
coinKeeper := bank.NewCoinKeeper(app.accountMapper)
app.Router().AddRoute("bank", bank.NewHandler(coinKeeper))
```

jaekwon commented on examples/basecoin/cmd/basecoind/main.go:
```diff
@@ -1,10 +1,13 @@
 
 	bapp := app.NewBasecoinApp("")
-	bapp.RunForever()
+	baseapp.RunForever(bapp)
```
bapp.RunForever()?

rigelrozanski replied:
This is the simplest way to achieve a reusable RunForever function (found in `baseapp/helpers.go`). Don't really see another easy to inherit from baseapp because `server.NewServer("0.0.0.0:46658", "socket", app)` takes the actual abci.Application as an input

ebuchman commented:
Closing in favour of https://github.com/cosmos/cosmos-sdk/pull/470
